src/mbgl/style/layer_impl.hpp

```diff
@@ -30,24 +27,17 @@ namespace style {
  * Members that are public in `FooLayer::Impl` are part of the internal API for "foo" layers.
  * Members that are private in `FooLayer::Impl` are internal to "foo" layers.
  */
-class Layer::Impl {
+class Layer::Impl : public EnableImmutableFromThis<Layer::Impl> {
```

Comment:
`EnableImmutableFromThis` is required only for `Source::Impl::createRenderSource` and `Layer::Impl::createRenderLayer`. It's probably a good idea to change how those operations are performed so that `EnableImmutableFromThis` isn't necessary. I see two options:

* Require passing an `Immutable<Impl>` self reference to those methods (3e7ce6ba27b31a73a9242e4e14e8eaacefe42a5e). I'm not crazy about this because it allows the possibility of passing some other reference that doesn't match `this`.
* Replace the virtual dispatch with a static dispatch -- switch statement or `Layer::accept`.